File: skills/current/market-hunter/lib/adapters/z2u.ts
import { Effect } from "effect";
import type { AdapterError, MarketplaceAdapter, RawMarketListing, SearchTarget } from "#models";
import { normalizeRawItems, type RawScrapedItem } from "./common.ts";

export class Z2uAdapter implements MarketplaceAdapter {
  readonly id = "z2u" as const;
  readonly displayName = "Z2U";
  readonly isEnabledByDefault = true;

  buildSearchTarget(query: string): SearchTarget {
    const encoded = encodeURIComponent(query.trim());
    return {
      marketplace: this.id,
      url: `https://www.z2u.com/search?q=${encoded}`,
      waitForMs: 2000,
      format: "json",
    };
  }

  parseListings(raw: unknown): Effect.Effect<readonly RawMarketListing[], AdapterError> {
    if (!raw) return Effect.succeed([]);

    if (typeof raw === "string") {
      const items: RawScrapedItem[] = [];
      const linkRegex = /\[([^\]]+)\]\((https:\/\/www\.z2u\.com\/product-[^)]+)\)/g;
      let match: RegExpExecArray | null;

      while ((match = linkRegex.exec(raw)) !== null) {
        const fullBlock = match[1] || "";
        const url = match[2];
        const priceMatch = /from\$([0-9]+(?:\.[0-9]+)?)/.exec(fullBlock) || /\$([0-9]+(?:\.[0-9]+)?)/.exec(fullBlock);
        const titleMatch = /^([^\\\[]+)/.exec(fullBlock);

        const title = titleMatch ? titleMatch[1]?.trim() : "";
        const price = priceMatch ? priceMatch[1] : "";

        if (title && price) {
          items.push({
            id: url ? url.split("/product-")[1]?.split("/")[0] || `z2u-${items.length + 1}` : undefined,
            title,
            url,
            priceUsd: price,
            sellerRating: 98,
          });
        }
      }

      return normalizeRawItems(items, this.id, 14);
    }

    if (typeof raw === "object") {
      const obj = raw as Record<string, unknown>;
      const rawItems = Array.isArray(obj["items"])
        ? (obj["items"] as readonly RawScrapedItem[])
        : Array.isArray(obj["products"])
          ? (obj["products"] as readonly RawScrapedItem[])
          : [];

      return normalizeRawItems(rawItems, this.id, 14);
    }

    return Effect.succeed([]);
  }
}

File: skills/current/market-hunter/lib/adapters/z2u.py
"""Z2U marketplace adapter (port of lib/adapters/z2u.ts)."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from adapters.common import RawScrapedItem, normalize_raw_items

if TYPE_CHECKING:
    from models import MarketplaceId, RawMarketListing, SearchTarget

_LINK_RE = re.compile(r"\[([^\]]+)\]\((https://www\.z2u\.com/product-[^)]+)\)")
_FROM_PRICE_RE = re.compile(r"from\$([0-9]+(?:\.[0-9]+)?)")
_PRICE_RE = re.compile(r"\$([0-9]+(?:\.[0-9]+)?)")
_TITLE_RE = re.compile(r"([^\\\[]+)")


class Z2uAdapter:
    """Z2U marketplace adapter."""

    id: MarketplaceId = "z2u"
    display_name = "Z2U"
    is_enabled_by_default = True

    def build_search_target(self, query: str) -> SearchTarget:
        """Build the Z2U search URL for a query."""
        encoded = quote(query.strip(), safe="!'()*")
        return {
            "marketplace": self.id,
            "url": f"https://www.z2u.com/search?q={encoded}",
            "waitForMs": 2000,
            "format": "json",
        }

    def parse_listings(self, raw: Any) -> list[RawMarketListing]:
        """Parse Z2U markdown or JSON scrape output into listings."""
        if not raw:
            return []

        if isinstance(raw, str):
            items: list[RawScrapedItem] = []
            for match in _LINK_RE.finditer(raw):
                full_block = match.group(1) or ""
                url = match.group(2)
                price_match = _FROM_PRICE_RE.search(full_block) or _PRICE_RE.search(
                    full_block
                )
                title_match = _TITLE_RE.match(full_block)

                title = title_match.group(1).strip() if title_match else ""
                price = price_match.group(1) if price_match else ""

                if title and price:
                    item: RawScrapedItem = {
                        "title": title,
                        "url": url,
                        "priceUsd": price,
                        "sellerRating": 98,
                    }
                    if url:
                        parts = url.split("/product-")
                        item_id = parts[1].split("/")[0] if len(parts) > 1 else ""
                        item["id"] = item_id or f"z2u-{len(items) + 1}"
                    items.append(item)

            return normalize_raw_items(items, self.id, 14)

        if isinstance(raw, dict):
            raw_items = raw.get("items")
            if not isinstance(raw_items, list):
                raw_items = raw.get("products")
            return normalize_raw_items(
                raw_items if isinstance(raw_items, list) else [],
                self.id,
                14,
            )

        return []
